hail/python/hailtop/batch/job.py

```diff
@@ -90,6 +95,16 @@ def __init__(self,
         self._valid: Set[_resource.Resource] = set()  # resources declared in the appropriate place
         self._dependencies: Set[Job] = set()
 
+        def safe_str(s):
+            new_s = ''
+            for c in s:
+                if c.isalnum() or c == '-':
+                    new_s += c
+                else:
+                    new_s += '_'
```

Comment:
Since strings in Python are immutable, this function isn't O(N). You can either use a [StringIO](https://docs.python.org/3/library/io.html#io.StringIO) to accumulate the characters and then call `getvalue` or you can create a list of characters and then call `''.join(ls)`